stellarobservatory/quorum_lachowski.py

```diff
@@ -122,19 +128,64 @@ def contract_to_maximal_quorum(nodes, slices_by_node):
     while True:
         filtered = set()
         for node in nodes:
-            if contains_quorum_slice(nodes, slices_by_node, node):
+            if contains_slice(nodes, slices_by_node, node):
                 filtered.add(node)
         if filtered in (nodes, {}):
             return filtered
         nodes = filtered
 
-def contains_quorum_slice(nodes_subset, slices, node):
-    """Check if for the given nodes and quorum slices there is a quorum slice
-    contained in the set of given nodes."""
+
+def is_quorum(slices, nodes_subset):
+    """
+    Check whether nodes_subset is a quorum in FBAS F (implicitly passed in via slices).
+    """
+    return all([
+        contains_slice(nodes_subset, slices, v)
+        for v in nodes_subset
+    ])
+
+
+def contains_slice(nodes_subset, slices, node):
+    """Check if for the given node quorum slices there is a quorum slice
+    contained in the subset of nodes.
+    Input: FBAS(V,S) implicitly passed in via slices; nodes_subset ⊆ V; node ∈ V
+    Output: whether node has a quorum slice contained in nodes_subset"""
     return any(quorum_slice.issubset(nodes_subset) for quorum_slice in slices[node])
 
+
 def next_split_node(nodes_subset, deps_by_node):
     """Choose the next split node to process: uniformly at random pick a node with max in-degree"""
     induced_subgraph = graph.get_induced_subgraph(deps_by_node, nodes_subset)
     indegrees = graph.get_indegrees(induced_subgraph)
     return max(indegrees.items(), key=operator.itemgetter(1))[0]
+
+
+def enumerate_quorums(slices_by_node):
+    """Enumerate all quorums of FBAS F (given by slices_by_node)."""
+    deps_by_node = get_dependencies_by_node(slices_by_node)
+    fbas_info = {
+        'deps_by_node': deps_by_node,
+        'slices_by_node': slices_by_node
+    }
+    all_nodes = set(slices_by_node.keys()) # all nodes need to be present as keys here
+    return traverse_quorums(fbas_info, set(), all_nodes)
+
+
+def traverse_quorums(fbas_info, committed, remaining):
+    """Given a FBAS F (by fbas_info) with set of nodes V
+    and given the sets: committed ⊆ V; R ⊆ V\\committed,
+    enumerate all quorums Q of F with committed ⊆ Q ⊆ committed ∪ remaining"""
+    if remaining == set():
+        if is_quorum(fbas_info['slices_by_node'], committed):
+            logging.debug("found quorum: %s", committed)
+            yield frozenset(committed)
+    else:
+        perimeter = committed.union(remaining)
+        greatest_q = greatest_quorum(perimeter, fbas_info['slices_by_node'])
+        if greatest_q == set() or not committed.issubset(greatest_q):
+            return
+        # v ← pick from R:
+        split = next_split_node(remaining, fbas_info['deps_by_node'])
+        remaining_without_split = remaining.difference({split})
+        yield from traverse_quorums(fbas_info, committed, remaining_without_split)
+        yield from traverse_quorums(fbas_info, committed.union({split}), remaining_without_split)
```

Comment:
This is beautiful! Very readable and almost line by line as @torstenstueber's algorithm. :clap: